vignettes/netCDF_Functionality.Rmd

```diff
@@ -0,0 +1,400 @@
+  crs ="+init=epsg:4326",
+  resolution = c(0.5, 0.5)
+)
+r[] <- seq_len(prod(dim(r)))
+xy <- raster::sampleRandom(r, size = 50, sp = TRUE)
+grid_cell_area <- calculate_cell_area(sites = xy, grid = r)
+
+calculate_nominal_resolution(
+  grid = r,
+  sites = xy,
+  cell_areas_km2 = grid_cell_area[, "km2"]
+)
+```
+
+# Output Files Structure Naming Conventions
+
+File structure should be organized by: variableName / allFilesPerVariable. Each file
+will contain variable output for one time period and/or climate model.
+
+The name for each file should be in the format as follows:
+**var_LorA_timestep_Project_SOILWAT2_GCM_RCP_variantLabel_gn_startT_endT.nc**
+
+where:
+
+* var – Shorthand name for variable that is contained in files (i.e. ppt for precipitation, WDD for wet degree days)
+* LorA – Land (L) or atmospheric (A) data
+* timestep – The time-step of the data contained (i.e. the units of the time dimension (i.e. year, month)).
+* Project – Project Name (i.e. SEUG or RangeResil, etc.)
+* SOILWAT2 – Keep as is. Denotes that SOILWAT2 was used to generate these results
+* GCM – Name of the GCM climate data used to drive the SOILWAT2 model (i.e. IPSL5, CCSM2-1)
+* RCP – Name of which GCM used to drive the GCM (RCP4.5, RCP8.5)
+* variantLabel –Variant of the GCM used (i.e. r1i1p1f1)
+* gn
+* startT – lower time bound or start of the time period contained. Units should be the date (YearMonthDay, XXXXYYZZ).
+* endT – upper time bound or end of the time period contained. Units should be the date (YearMonthDay, XXXXYYZZ).
+
+# Create netCDF Function and Examples
+
+The functions create_empty_netCDF_file and populate_netcdf_from_array will create and populate
+a netCDF from an array. The user defines the dimensions, attribtues, and locations in the functions'
+arguments.
+
+The two functions work in conjuction with one another. First, an empty netcdf is created.
+Then, it is filled with data.
+
+## Example 1 - Create a 3 dimension netCDF (lat, long, time) from gridded data
+```{r, 3dGridExample, results = 'hide'}
+    ## locations
+    crs <- '+init=epsg:4326' # the CRS
+
+    outFileName <- 'dummynetcdf.nc'
+
+    # create
+    create_empty_netCDF_file(
+      data = data1d,
+      has_T_timeAxis = TRUE,
+      has_Z_verticalAxis = FALSE,
+      time_bounds = annual_time_bounds,
+      vert_bounds = NULL,
+      var_attributes = var_attributes,
+      time_attributes  = time_attributes,
+      vertical_attributes = NULL,
+      global_attributes = global_attributes,
+      isGridded = TRUE,
+      locations = locationsGrid,
+      crs = crs,
+      file = outFileName,
+      force_v4 = TRUE,
+      overwrite = TRUE
+    )
+
+    # populate
+    populate_netcdf_from_array(
+      file = outFileName,
+      data = data1d,
+      var_names <- var_attributes$name,
+      has_T_timeAxis = TRUE,
+      has_Z_verticalAxis = FALSE,
+      isGridded = TRUE,
+      locations = locationsGrid,
+      crs = crs,
+      force_v4 = TRUE
+    )
+
```

Comment:
I would like to see a demonstration for each example how to read and recreate the inputs successfully, e.g., something like
```
all.equal(read_data(...), data1d)

tmp <- read_coordinates(...)
all.equal(tmp[["coords_spatial"]], locationsGrid)
all.equal(tmp[["coords_vertical"]], vertical_attributes[["vals"]])
all.equal(tmp[["coords_time"]], time_attributes[["vals"]])
```

Reply:
added

Reply:
Well I made sure all well values were equal. Not other attributes.

Reply:
Nice!